--- a/libsofia-sip-ua/nua/nua_client.c	2021-02-23 17:53:17.000000000 +0100
+++ b/libsofia-sip-ua/nua/nua_client.c	2021-07-11 15:04:52.270162344 +0200
@@ -880,11 +880,6 @@
     if (ds->ds_ltarget && !cr->cr_has_contact)
       sip_add_dup(msg, sip, (sip_header_t *)ds->ds_ltarget);
 
-	/*
-	  FS-4102 
-	  It was decided to comment out this code because it does not appear to make sense
-	  Dec 22, 2016
-
     if (nua_registration_add_contact_to_request(nh, msg, sip,
 						cr->cr_contactize &&
 						!cr->cr_has_contact &&
@@ -893,8 +888,6 @@
       msg_destroy(msg);
       return -1;
     }
-	*/
-
   }
 
   cr->cr_wait_for_cred = 0;
